### DISCUSSION TITLE
[Help] "use client" Could not get process.env.NEXT_PUBLIC_APP_ENV

### DISCUSSION THREAD

**Discussion by darkdx** [vercel/next.js#51615] [Help] [answered]
### Summary

When I use "use client", I cannot get process.env.NEXT_PUBLIC_APP_ENV.

<img width="949" alt="image" src="https://github.com/vercel/next.js/assets/88136242/75d1089c-0c44-4209-b6bf-f1a3576a1dfa">

```json
// [packeage.json] 

  "dependencies": {
    "@headlessui/react": "^1.7.15",
    "@heroicons/react": "^2.0.18",
    "@rainbow-me/rainbowkit": "^1.0.2",
    "@types/node": "20.3.1",
    "@types/react": "18.2.13",
    "@types/react-dom": "18.2.6",
    "autoprefixer": "10.4.14",
    "clsx": "^1.2.1",
    "eslint": "8.43.0",
    "eslint-config-next": "13.4.6",
    "next": "13.4.6",
    "next-intl": "^2.15.1",
    "postcss": "8.4.24",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "tailwind-merge": "^1.13.2",
    "tailwindcss": "3.3.2",
    "typescript": "5.1.3",
    "viem": "^1.1.0",
    "wagmi": "^1.2.0"
  }

```

**Comment by darkdx** [comment]
```tsx
"use client"

export default function About() {
  console.log('process', process.env?.NEXT_PUBLIC_APP_ENV)

  return (
    <main className="flex min-h-screen flex-col items-center justify-between p-24">
      about
    </main>
  );
}



```

**Comment by hesalx** [comment, marked as answer]
Try `process.env.NEXT_PUBLIC_APP_ENV` instead of `process.env?.NEXT_PUBLIC_APP_ENV`. It is looking for specific syntax because it is statically replaced during the build. You don't have to worry about `?.` in this case.

**Reply by darkdx** [reply]
You are absolutely right, it has solved my problem, thank you very much, have a good day.

**Comment by rasif-sahl** [comment]
1. Try `process.env.NEXT_PUBLIC_APP_ENV`
2. Make sure that the environment variable NEXT_PUBLIC_APP_ENV is defined in your `.env.local` file or any other environment file that Next.js recognizes. The .env.local file should be located in the root directory of your project.
3. Restart your development server after making any changes to the environment variables.
4. Confirm that you are using the correct syntax to access the environment variable in your code. It should be accessed as process.env.NEXT_PUBLIC_APP_ENV (note the process.env prefix).
5. run `console.log(process.env.NEXT_PUBLIC_APP_ENV)` and see whether you can access the env variable

**Reply by darkdx** [reply]
You are absolutely right, it has solved my problem, thank you very much, have a good day.

**Reply by rasif-sahl** [reply]
Great

**Comment by icyJoseph** [comment]
Hi,

When you do `process.env?`, it affects the compilation of process.env - note that this doesn't exist in the browser window natively, and it is instead added by the bundler.

```js
function About() {
    var _process_env;
    console.log("process", (_process_env = process.env) === null || _process_env === void 0 ? void 0 : _process_env.NEXT_PUBLIC_APP_ENV);
    return /*#__PURE__*/ (0,react_jsx_dev_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxDEV)("main", {
        className: "flex min-h-screen flex-col items-center justify-between p-24",
        children: "about"
    }, void 0, false, {
        fileName: "/Users/joseph/dev-exp/app-dir-tester/src/app/foo/page.tsx",
        lineNumber: 7,
        columnNumber: 5
    }, this);
}
```

This is what your component resolves too when you use `process.env?` - and this is without the optional chaining:

```js
function About() {
    console.log("process", process.env.NEXT_PUBLIC_APP_ENV);
    return /*#__PURE__*/ (0,react_jsx_dev_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxDEV)("main", {
        className: "flex min-h-screen flex-col items-center justify-between p-24",
        children: "about"
    }, void 0, false, {
        fileName: "/Users/joseph/dev-exp/app-dir-tester/src/app/foo/page.tsx",
        lineNumber: 7,
        columnNumber: 5
    }, this);
}
```

This is to say that your approach also fails in Pages Dir!

This is the same reason why, https://nextjs.org/docs/app/api-reference/next-config-js/env:

> Trying to destructure process.env variables won't work due to the nature of webpack [DefinePlugin](https://webpack.js.org/plugins/define-plugin/).

**Reply by darkdx** [reply]
You are absolutely right, it has solved my problem, thank you very much, have a good day.
